### PR TITLE
Add SOURCE keyword

### PR DESCRIPTION
**Pull request by totto82**
Add keyword for setting a source term in the simulator. 
```
SOURCE
-- i j k component massrate 
 1 1 1  GAS -0.1 /
```
This is needed to support the 11th SPE Comparative Solutions Project.

### CODE REVIEW

File: opm/input/eclipse/Schedule/SourceProp.hpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ */
+
+#ifndef OPM_SOURCE_PROP_HPP
+#define OPM_SOURCE_PROP_HPP
+
+#include <vector>
+#include <cstddef>
+#include <optional>
+
+#include <opm/input/eclipse/EclipseState/Grid/GridDims.hpp>
```
Comment (bska) [inline opm/input/eclipse/Schedule/SourceProp.hpp:28]: Why do you need `GridDims.hpp` here?

File: opm/input/eclipse/Schedule/SourceProp.hpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     SOLVENT,
+     POLYMER,
+     NONE
+};
+
+class SourceProp {
+public:
+
+    struct SourceCell {
+        int i, j, k;
```
Comment (bska) [inline opm/input/eclipse/Schedule/SourceProp.hpp:48]: Do you really need to refer to the individual `i`, `j`, and `k` coordinates?  If not, maybe we could collect these into a `std::array<int,3>` instead?  That would also simplify the comparison operators.

**Review by bska — COMMENTED** [review]
I think we should attempt to stick to using `ptr_member` types where possible in the `ScheduleState`.  Those are really the way the Schedule object is **supposed** to fit together.  Other than that, I just have a couple of minor comments.

File: opm/input/eclipse/Schedule/ScheduleState.hpp [hunk 3]
```diff
@@ -480,6 +481,7 @@ namespace Opm {
         // constant flux aquifers
         std::unordered_map<int, SingleAquiferFlux> aqufluxs;
         BCProp bcprop;
+        SourceProp sourceprop;
```
[the remark this answers is not in the document]
Comment (totto82) [reply opm/input/eclipse/Schedule/ScheduleState.hpp:484]: I tried to change to `ptr_member` types but I got lost. How is ptr_member suppose to work?

[the remark this answers is not in the document]
Comment (bska) [reply opm/input/eclipse/Schedule/ScheduleState.hpp:484]: > How is `ptr_member` suppose to work?

You create a default-constructed object in [`Schedule::create_first()`](https://github.com/OPM/opm-common/blob/856506b2474043bf61fc194efd86578101346d0b/src/opm/input/eclipse/Schedule/Schedule.cpp#L2183) and then the `ScheduleState` copy-constructor ensures that this default-constructed object gets shared with all other `ScheduleState` instances until the keyword is encountered in the SCHEDULE section.  In the keyword handler you then generally proceed as the [handler for `GUIDERAT`](https://github.com/OPM/opm-common/blob/856506b2474043bf61fc194efd86578101346d0b/src/opm/input/eclipse/Schedule/Group/GuideRateKeywordHandlers.cpp#L39)

1. You [process/internalise](https://github.com/OPM/opm-common/blob/856506b2474043bf61fc194efd86578101346d0b/src/opm/input/eclipse/Schedule/Group/GuideRateKeywordHandlers.cpp#L43-L55) the records of the keywords into an independent (`SourceProp`) object
2. You compare that object to the [current keyword object](https://github.com/OPM/opm-common/blob/856506b2474043bf61fc194efd86578101346d0b/src/opm/input/eclipse/Schedule/Group/GuideRateKeywordHandlers.cpp#L56)
3. If they differ, then you [`update()`](https://github.com/OPM/opm-common/blob/856506b2474043bf61fc194efd86578101346d0b/src/opm/input/eclipse/Schedule/Group/GuideRateKeywordHandlers.cpp#L58) the current object which will then be independent of the `SourceProp` object at earlier times

The new current object will be shared among all following `ScheduleState` objects until the next `SOURCE` keyword is encountered, at which point the above process repeats.

At the client side, you generally write code like (with `sources` being my made-up name for what's currently `ScheduleState::sourceprop`
```C++
const auto& src = schedule[step].sources();
use(src);
```
to incorporate `src` into whatever context you need, e.g., the material balance equations.

[the remark this answers is not in the document]
Comment (totto82) [reply opm/input/eclipse/Schedule/ScheduleState.hpp]: Thanks. I have updated accordingly. I also changed from sourceprop to source

**Comment by totto82** [comment]
jenkins build this opm-simulators=5050 please

**Comment by totto82** [comment]
jenkins build this opm-simulators=5050 please

**Comment by totto82** [comment]
jenkins build this opm-simulators=5050 please

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp [hunk 4]
```diff
@@ -96,6 +96,15 @@ void handleBCProp(HandlerContext& handlerContext)
     }
 }
 
+void handleSource(HandlerContext& handlerContext)
+{
+    auto new_source = Source();
+    for (const auto& record : handlerContext.keyword) {
+        new_source.updateSource(record);
+    }
+    handlerContext.state().source.update(std::move(new_source));
```
Comment (bska) [inline src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:105]: This definition (especially `new_source = Source{}`) will replace all existing source terms with the contents of the new `SOURCE` keyword if the keyword is entered multiple times.  Is that the intended/desired semantics of this keyword?

**Review by bska — COMMENTED** [review]
Thanks a lot for switching to using `ptr_member<>`&ndash;that's greatly appreciated.  The only **potential** problem I see in the current implementation is that if the `SOURCE` keyword is entered multiple times then earlier entries will be completely replaced by later entries.

On the other hand, if that's the desired semantics, then I'm happy with the current version.

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp (hunk 4 above)
Comment (totto82) [reply src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:105]: That was not intended. The intention (also reflected in one of the test cases) was that old source terms should be kept until they are overwritten by a source with the same i.j,k index. Strangely the test pass. 

```
SCHEDULE
SOURCE
 1 1 1 GAS 0.01 /
 1 1 1 WATER 0.01 /
/
DATES             -- 1
 10  'JUN'  2007 /
/
SOURCE
 1 1 1 GAS 0.0 /
 1 1 2 WATER 0.02 /
/
```
Still gives 0.01 of water for cell 1 1 1.

File: src/opm/input/eclipse/share/keywords/900_OPM/S/SOURCE [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    },
+    {
+       "name": "COMPONENT",
+       "value_type": "STRING",
+       "default": "NONE"
+    },
+    {
+       "name": "RATE",
+       "value_type": "DOUBLE",
+       "dimension": "Mass/Time",
```
[the remark this answers is not in the document]
Comment (totto82) [reply src/opm/input/eclipse/share/keywords/900_OPM/S/SOURCE:27]: 0.1 is just a random number. Note that I have changed the sign. I use plus for source and minus for sink.

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp (hunk 4 above)
Comment (totto82) [reply src/opm/input/eclipse/Schedule/KeywordHandlers.cpp]: I have updated the code and added a test for Schedule. Thanks for noticing.

**Comment by totto82** [comment]
jenkins build this opm-simulators=5050 please

**Comment by totto82** [comment]
> If you'd like to make this extension then I'll hold off

Sound like a good idea. Lets cross that bridge when we get there. Thanks for the review.

**Comment by atgeirr** [comment]
Reviewed, all green, merging.

**Pull request merged**